cur-lib/cur/stdlib/sugar.rkt

```diff
@@ -184,53 +187,86 @@
+        (~and decls-y [y (~datum :) ty_in2]) ...
         (~datum ->) ty_out
-        [pat ... (~datum =>) body] ...)
-     #:with (decls-x ...) (for/list ([x (attribute x)]
-                                     [t (attribute ty_in1)])
-                            #`(#,x : #,t))
-     #:with (decls-y ...) (for/list ([y (attribute y)]
-                                     [t (attribute ty_in2)])
-                            #`(#,y : #,t))
-     #:with (pat-bodies ...) (for/list ([pats (attribute pat)]
-                                        [body (attribute body)])
-                               #`(#,pats => #,body))
+        (~and clauses
+              [pat ... (~datum =>) body]) ...
+        (~optional
+         (~seq #:type-aliases
+               ([ty-in-aliases (~datum =) ty-out-aliases
+                               (~optional
+                                ty-out-implicit-counts:exact-nonnegative-integer
+                                ;; TODO PR103: Why is this a syntax-object, while others
+                                ;; are not?
+                                #:defaults ([ty-out-implicit-counts #'0]))] ...))
+         #:defaults ([(ty-in-aliases 1) '()]
+                     [(ty-out-aliases 1) '()]
+                     [(ty-out-implicit-counts 1) '()])))
+     ; do totality checking
+     #:with (temporaries ...) (for/list ([ty (attribute ty-to-match)]
+                                         [idx (in-naturals)])
+                                (format-id ty "~a" (generate-temporary (string->symbol (format "match-~a-" (add1 idx)))) #:source ty))
+     #:do [(total? (list (attribute temporaries) (attribute clauses))
+                   #:aliases (map list
+                                  (attribute ty-in-aliases)
+                                  (attribute ty-out-aliases)
+                                  (map syntax->datum (attribute ty-out-implicit-counts)))
+                   #:env (reverse (append
+                                   (map cons (attribute x) (attribute ty_in1))
+                                   (map cons (attribute temporaries) (attribute ty-to-match))
+                                   (map cons (attribute y) (attribute ty_in2)))))]
      ; Explicitly mark different parameters as the decreasing argument for
      ; termination checking. the last iteration is guaranteed to either be
      ; successful or another exception that isn't exn:fail:recur.
+     ;; TODO PR103: Should probably define a phase-1 function instead of
+     ;; local-expanding define/rec/match^.
      (if (zero? (stx-length #'(ty-to-match ...)))
-         (local-expand #`(define/rec/match^ name : decls-x ... ty-to-match ... decls-y ... -> ty_out pat-bodies ...) 'top-level null)
+         (local-expand
+          #`(define/rec/match^ name : decls-x ... ty-to-match ...
+              decls-y ... -> ty_out clauses ...)
+          'top-level null)
+         ;; TODO PR103: What happens if all fail? Surely the function wouldn't be
```

Comment:
You're right, it would be better to have an explicit failure maybe. I think the reasoning was that if another error occurred, then it would have raised an exception already. Reaching the end here would be an error in implementation of the termination checker. (see comment on line 233)